src/utils/atomic_write.py

```
@@ -0,0 +1,275 @@
+    encoding: str | None = None,
+) -> None:
+    """Atomically write via temp file + fsync + ``os.replace``.
+
+    Creates a temp file in ``target.parent``, runs ``write_fn(handle)``,
+    fsyncs, atomically replaces ``target``. Cleans up the temp on any
+    exception (writer error, KeyboardInterrupt, failed ``os.replace``).
+    """
+    # ``NamedTemporaryFile(delete=False)`` leaves the temp around so we
+    # can ``os.replace`` it into place. ``dir=target.parent`` is
+    # essential — cross-device renames raise ``EXDEV``.
+    tmp_file = tempfile.NamedTemporaryFile(
```

Ruff (0.15.11)

[warning] Use a context manager for opening files

(SIM115)